spi/w5500_spi/loopback.c

```diff
@@ -0,0 +1,225 @@
+#include <stdio.h>
+#include "loopback.h"
+#include "socket.h"
+#include "wizchip_conf.h"
+
+#if LOOPBACK_MODE == LOOPBACK_MAIN_NOBLCOK
+
+int32_t loopback_tcps(uint8_t sn, uint8_t* buf, uint16_t port)
+{
+   int32_t ret;
+   uint16_t size = 0, sentsize=0;
+
+#ifdef _LOOPBACK_DEBUG_
+   uint8_t destip[4];
+   uint16_t destport;
+#endif
+
+   switch(getSn_SR(sn))
+   {
+      case SOCK_ESTABLISHED :
+         if(getSn_IR(sn) & Sn_IR_CON)
+         {
+#ifdef _LOOPBACK_DEBUG_
+			   getSn_DIPR(sn, destip);
+			   destport = getSn_DPORT(sn);
+
+			   printf("%d:Connected - %d.%d.%d.%d : %d\r\n",sn, destip[0], destip[1], destip[2], destip[3], destport);
+#endif
+            setSn_IR(sn,Sn_IR_CON);
+         }
+         if((size = getSn_RX_RSR(sn)) > 0) // Don't need to check SOCKERR_BUSY because it doesn't not occur.
+         {
+			   if(size > DATA_BUF_SIZE) size = DATA_BUF_SIZE;
+			   ret = recv(sn, buf, size);
+
+			   if(ret <= 0) return ret;      // check SOCKERR_BUSY & SOCKERR_XXX. For showing the occurrence of SOCKERR_BUSY.
+			   size = (uint16_t) ret;
+			   sentsize = 0;
+
+			   while(size != sentsize)
+			   {
+				   ret = send(sn, buf+sentsize, size-sentsize);
+				   if(ret < 0)
+				   {
+					   close(sn);
+					   return ret;
+				   }
+				   sentsize += ret; // Don't care SOCKERR_BUSY, because it is zero.
+			   }
+         }
+         break;
+      case SOCK_CLOSE_WAIT :
+#ifdef _LOOPBACK_DEBUG_
```

Comment:
Really don't like the continual DEBUG ifdef's. Looks messy. This is example code. Either we remove all the debug statements, and leave them in as comments in the code, or remove the ifdefs.